onnx/defs/shape_inference.h

```diff
@@ -18,18 +18,22 @@ namespace ONNX_NAMESPACE {
 
 using Dim = TensorShapeProto_Dimension;
 
+enum InferenceErrorMode : uint8_t {
```

Comment:
nit: I think we could easily use `uint32_t`. Cost is not significant, and may help us in long run.

Reply:
I also wonder whether we need to expose anything like this externally? Who are the intended users and what will they use it for? May be it is better to focus on the intended use, which is internal ... specifically, we want to be able to propagate context information so that we know whether we are processing a top-level node, or a node contained inside a conditional (or loop? not sure about that). If it is a node inside a conditional, we want to tolerate some errors, since the node may never execute. So, I wonder if we really need this enumeration or expose it externally.

Reply:
Another concern here: the checker throws [ValidationError](https://github.com/onnx/onnx/blob/403eb3ecfb022c33ba2164595dc1d0a480a3544a/onnx/checker.h#L22). For the example discussed above, it may be necessary to catch/handle checker-errors as well, since a missing attribute might show up as a checker error.

Reply:
I think it might be helpful to validate ONNX with this PR: https://github.com/onnx/onnx/pull/5488 (with converters etc.) to see what failures we run into, if any. And create some test-cases with if-then-else that we would like to successfully pass the checker, and approach this top-down, to make sure we are able to do what we want.

Reply:
the enum is exposed via: infer_shapes, infer_shapes_path. It originally used a binary strict_mode, which can be replaced with IgnoreInferenceError or other enum values. In that sense, we can keep the 2 interfaces as it was so that the enum does not have to be visible from python. For checker, there is an option to run shape inference. In this case, the shape inference error mode is set to be FailAnyInferenceError. This means shape inference specific exception is thrown. As an example: https://github.com/onnx/onnx/blob/abf0fa73a313b97f9979ba25f5fc97a0a0c7fab5/onnx/test/checker_test.py#L499